yuyo/components.py

```diff
@@ -1333,6 +1335,102 @@ async def execute(
 """Alias for `WaitForExecutor`."""
 
 
+class ComponentStream(AbstractComponentExecutor):
+    __slots__ = ("_authors", "_ephemeral_default", "_finished", "_future", "_made_at", "_max_backlog", "_timeout")
+
+    def __init__(
+        self,
+        *,
+        authors: typing.Optional[typing.Iterable[hikari.SnowflakeishOr[hikari.User]]],
+        timeout: datetime.timedelta,
+        ephemeral_default: bool = False,
+        max_backlog: int = 5,
+    ) -> None:
+        self._authors = set(map(hikari.Snowflake, authors)) if authors else None
+        self._ephemeral_default = ephemeral_default
+        self._finished = False
+        self._future: typing.Optional[asyncio.Future[ComponentContext]] = None
+        self._made_at: typing.Optional[datetime.datetime] = None
+        self._max_backlog = max_backlog
+        self._queue: typing.Optional[asyncio.Queue[ComponentContext]] = None
+        self._timeout = timeout
+
+    @property
+    def custom_ids(self) -> typing.Collection[str]:
+        # <<inherited docstring from AbstractComponentExecutor>>.
+        return []
+
+    @property
+    def has_expired(self) -> bool:
+        # <<inherited docstring from AbstractComponentExecutor>>.
+        return bool(
+            self._finished
+            or self._made_at
+            and self._timeout < datetime.datetime.now(tz=datetime.timezone.utc) - self._made_at
+        )
+
+    def __enter__(self: _ComponentStreamT) -> _ComponentStreamT:
+        self.open()
+        return self
+
+    def __exit__(
+        self: _ComponentStreamT,
+        exc_type: typing.Optional[typing.Type[BaseException]],
+        exc: typing.Optional[BaseException],
+        exc_traceback: typing.Optional[types.TracebackType],
+    ) -> _ComponentStreamT:
+        self.close()
+        return self
+
+    def open(self) -> None:
+        if self._queue is not None:
+            raise RuntimeError("Stream is already active")
```

Comment:
If its already open just return instead of raising so they can await it while it's already open